hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java

```diff
@@ -1659,6 +1662,43 @@ public RegionLoad createRegionLoad(final String encodedRegionName) throws IOExce
     return r != null ? createRegionLoad(r, null, null) : null;
   }
 
+  public Map<String, Integer> uncacheStaleBlocks() {
+    Map<String, Pair<String, Long>> fullyCachedFiles =
+      this.getBlockCache().flatMap(BlockCache::getFullyCachedFiles).orElse(Collections.emptyMap());
+    Map<String, Integer> evictedFilesWithStaleBlocks = new ConcurrentHashMap<>();
+
+    ExecutorService executor = Executors.newFixedThreadPool(6);
+
+    List<Callable<Void>> tasks = new ArrayList<>();
+
+    fullyCachedFiles.forEach((fileName, value) -> {
+      Callable<Void> task = () -> {
+        HRegion regionOnServer = getRegion(value.getFirst());
+        int blocksEvicted = (regionOnServer == null || !regionOnServer.isAvailable())
+          ? this.getBlockCache().get().evictBlocksByHfileName(fileName)
+          : 0;
+        evictedFilesWithStaleBlocks.put(fileName, blocksEvicted);
+        LOG.info(
+          "Uncached {} blocks belonging to the file {} as the region {} "
+            + "is not served by the region server {} anymore.",
+          blocksEvicted, fileName, value.getFirst(), this.getServerName());
+        return null;
+      };
+      tasks.add(task);
```

Comment:
We should do a single task per call to the method, not a task per file, that would create yet another collection in memory with as many objects as the total files cached. On large caches, that would be impacting.